fuel-core-interfaces/src/p2p.rs

```diff
@@ -40,9 +72,48 @@ pub enum P2pRequestEvent {
     BroadcastConsensusVote {
         vote: Arc<ConsensusVote>,
     },
+    GossipsubMessageReport {
+        message: GossipsubMessageInfo,
+        acceptance: GossipsubMessageAcceptance,
+    },
     Stop,
 }
 
+#[derive(Debug, Clone)]
+pub struct GossipData<T> {
+    pub data: Option<T>,
+    pub peer_id: Vec<u8>,
+    pub message_id: Vec<u8>,
+}
+
+pub type ConsensusGossipData = GossipData<ConsensusBroadcast>;
+pub type TransactionGossipData = GossipData<TransactionBroadcast>;
+pub type BlockGossipData = GossipData<BlockBroadcast>;
+
+impl<T> GossipData<T> {
+    pub fn new(
+        data: T,
+        peer_id: impl Into<Vec<u8>>,
+        message_id: impl Into<Vec<u8>>,
+    ) -> Self {
+        Self {
+            data: Some(data),
+            peer_id: peer_id.into(),
+            message_id: message_id.into(),
+        }
+    }
+}
+
+pub trait NetworkData<T>: Debug + Send {
+    fn take_data(&mut self) -> Option<T>;
+}
```

Comment:
```suggestion
pub trait NetworkData<T>: Debug + Send {
    type Metadata;
    fn take_data(self) -> (Self::Metadata, T);
}
```
I'm thinking it would be better to actually split the type so that we don't have to handle the option?

Reply:
hmm maybe we can defer this to a followup task? Ideally in a way that wouldn't require users of this trait to know or care about the concrete type for metadata.

Reply:
I did raise this concern in the architecture PR but was told it was an implementation detail. Happy to do a follow up though

Reply:
yeah I think it's just a matter of having a fully fleshed out proposal of this actually working